README
NAME
    warnings::MaybeFatal - make warnings FATAL at compile-time only

SYNOPSIS
       use strict;
       use warnings qw(all);
       use warnings::MaybeFatal;
   
       # Use of uninitialized value.
       # Run-time warning, so this is non-fatal.
       print join(undef, "a", "b");
   
       # Useless use of constant in void context.
       # Compile-time warning, so this is fatalized.
       "Hello world"; 1;

DESCRIPTION
    Because it's kind of annoying if a warning stops your program from being
    compiled, but it's *really* annoying if it breaks your program part way
    through actually executing.

    This lexically scoped pragma will make all warnings (including custom
    warnings emitted with the `warn` keyword) FATAL during compile time. It
    does not enable or disable any warnings in its own right. It just makes
    any warnings that happen to be enabled FATAL during the compile.

    (Note that the compile phase and execute phase are not as cleanly divided
    in Perl as they are in, say, C. If module X loads module Y at run-time,
    then module Y's compile time happens during module X's run-time. In this
    situation, a warning that is triggered while compiling Y will be FATAL,
    even though from module X's perspective, this is at run-time.)

BUGS
    Please report any bugs to
    <http://rt.cpan.org/Dist/Display.html?Queue=warnings-MaybeFatal>.

SEE ALSO
    warnings.

AUTHOR
    Toby Inkster <tobyink@cpan.org>.

COPYRIGHT AND LICENCE
    This software is copyright (c) 2014 by Toby Inkster.

    This is free software; you can redistribute it and/or modify it under the
    same terms as the Perl 5 programming language system itself.

DISCLAIMER OF WARRANTIES
    THIS PACKAGE IS PROVIDED "AS IS" AND WITHOUT ANY EXPRESS OR IMPLIED
    WARRANTIES, INCLUDING, WITHOUT LIMITATION, THE IMPLIED WARRANTIES OF
    MERCHANTIBILITY AND FITNESS FOR A PARTICULAR PURPOSE.



Changes
warnings-MaybeFatal
===================

Created:      2014-01-30
Home page:    <https://metacpan.org/release/warnings-MaybeFatal>
Bug tracker:  <http://rt.cpan.org/Dist/Display.html?Queue=warnings-MaybeFatal>
Maintainer:   Toby Inkster (TOBYINK) <tobyink@cpan.org>

0.004	2014-02-06

 - warnings::MaybeFatal no longer dies as soon as a warning is issued.
   Instead, it allows the code to keep compiling, and only dies at the end
   of the compile if there were any warnings issued.
   Dagfinn Ilmari Mannsåker++
   Damien Krotkine++
   <http://purl.org/NET/cpan-uri/rt/ticket/92718>

0.003	2014-02-05

 [ Test Suite ]
 - Test that a module using warnings::MaybeFatal is able to load another
   module using warnings::MaybeFatal. (This test fails on Perl < 5.8.4.)
   Graham Knop++

 [ Other ]
 - Raise Perl version dependency from 5.8.1 to 5.8.4.

0.002	2014-02-05

 - Depend on Lexical::SealRequireHints for Perl 5.8.x and 5.10.x.
   Graham Knop++
   <http://purl.org/NET/cpan-uri/rt/ticket/92719>
 - No longer depend on Devel::Pragma for Perl 5.8.x.
   Graham Knop++
   <http://purl.org/NET/cpan-uri/rt/ticket/92719>
 - Reduce Perl version dependency from 5.8.3 to 5.8.1.

0.001	2014-02-01	Initial release


Makefile.PL
use strict;
use ExtUtils::MakeMaker 6.17;

my $EUMM = eval( $ExtUtils::MakeMaker::VERSION );

my $meta = {
  "abstract"       => "make warnings FATAL at compile-time only",
  "author"         => ["Toby Inkster (TOBYINK) <tobyink\@cpan.org>"],
  "dynamic_config" => 1,
  "generated_by"   => "Dist::Inkt::Profile::TOBYINK version 0.014, CPAN::Meta::Converter version 2.120921",
  "keywords"       => [],
  "license"        => ["perl_5"],
  "meta-spec"      => {
                        url => "http://search.cpan.org/perldoc?CPAN::Meta::Spec",
                        version => 2,
                      },
  "name"           => "warnings-MaybeFatal",
  "no_index"       => { directory => ["eg", "examples", "inc", "t", "xt"] },
  "prereqs"        => {
                        configure => { requires => { "ExtUtils::MakeMaker" => 6.17 } },
                        develop   => { recommends => { "Dist::Inkt" => 0.001 } },
                        runtime   => { requires => { "B::Hooks::EndOfScope" => 0, "perl" => 5.008004 } },
                        test      => {
                                       requires => { "Test::Fatal" => 0, "Test::More" => 0.96, "Test::Warnings" => 0 },
                                     },
                      },
  "provides"       => {
                        "warnings::MaybeFatal" => { file => "lib/warnings/MaybeFatal.pm", version => 0.004 },
                      },
  "release_status" => "stable",
  "resources"      => {
                        bugtracker   => {
                                          web => "http://rt.cpan.org/Dist/Display.html?Queue=warnings-MaybeFatal",
                                        },
                        homepage     => "https://metacpan.org/release/warnings-MaybeFatal",
                        license      => ["http://dev.perl.org/licenses/"],
                        repository   => {
                                          type => "git",
                                          web  => "https://github.com/tobyink/p5-warnings-maybefatal",
                                        },
                        X_identifier => "http://purl.org/NET/cpan-uri/dist/warnings-MaybeFatal/project",
                      },
  "version"        => 0.004,
};

my %dynamic_config;
do {
if ( $] < 5.012 )
{
	$meta->{prereqs}{runtime}{requires}{'Lexical::SealRequireHints'} = '0.007';
}

};

my %WriteMakefileArgs = (
	ABSTRACT   => $meta->{abstract},
	AUTHOR     => ($EUMM >= 6.5702 ? $meta->{author} : $meta->{author}[0]),
	DISTNAME   => $meta->{name},
	VERSION    => $meta->{version},
	EXE_FILES  => [ map $_->{file}, values %{ $meta->{x_provides_scripts} || {} } ],
	NAME       => do { my $n = $meta->{name}; $n =~ s/-/::/g; $n },
	test       => { TESTS => "t/*.t" },
	%dynamic_config,
);

$WriteMakefileArgs{LICENSE} = $meta->{license}[0] if $EUMM >= 6.3001;

sub deps
{
	my %r;
	for my $stage (@_)
	{
		for my $dep (keys %{$meta->{prereqs}{$stage}{requires}})
		{
			next if $dep eq 'perl';
			my $ver = $meta->{prereqs}{$stage}{requires}{$dep};
			$r{$dep} = $ver if !exists($r{$dep}) || $ver >= $r{$dep};
		}
	}
	\%r;
}

my ($build_requires, $configure_requires, $runtime_requires, $test_requires);
if ($EUMM >= 6.6303)
{
	$WriteMakefileArgs{BUILD_REQUIRES}     ||= deps('build');
	$WriteMakefileArgs{CONFIGURE_REQUIRES} ||= deps('configure');
	$WriteMakefileArgs{TEST_REQUIRES}      ||= deps('test');
	$WriteMakefileArgs{PREREQ_PM}          ||= deps('runtime');
}
elsif ($EUMM >= 6.5503)
{
	$WriteMakefileArgs{BUILD_REQUIRES}     ||= deps('build', 'test');
	$WriteMakefileArgs{CONFIGURE_REQUIRES} ||= deps('configure');
	$WriteMakefileArgs{PREREQ_PM}          ||= deps('runtime');	
}
elsif ($EUMM >= 6.52)
{
	$WriteMakefileArgs{CONFIGURE_REQUIRES} ||= deps('configure');
	$WriteMakefileArgs{PREREQ_PM}          ||= deps('runtime', 'build', 'test');	
}
else
{
	$WriteMakefileArgs{PREREQ_PM}          ||= deps('configure', 'build', 'test', 'runtime');	
}

{
	my ($minperl) = reverse sort(
		grep defined && /^[0-9]+(\.[0-9]+)?$/,
		map $meta->{prereqs}{$_}{requires}{perl},
		qw( configure build runtime )
	);
	
	if (defined($minperl))
	{
		die "Installing $meta->{name} requires Perl >= $minperl"
			unless $] >= $minperl;
		
		$WriteMakefileArgs{MIN_PERL_VERSION} ||= $minperl
			if $EUMM >= 6.48;
	}
}

sub FixMakefile
{
	return unless -d 'inc';
	my $file = shift;
	
	local *MAKEFILE;
	open MAKEFILE, "< $file" or die "FixMakefile: Couldn't open $file: $!; bailing out";
	my $makefile = do { local $/; <MAKEFILE> };
	close MAKEFILE or die $!;
	
	$makefile =~ s/\b(test_harness\(\$\(TEST_VERBOSE\), )/$1'inc', /;
	$makefile =~ s/( -I\$\(INST_ARCHLIB\))/ -Iinc$1/g;
	$makefile =~ s/( "-I\$\(INST_LIB\)")/ "-Iinc"$1/g;
	$makefile =~ s/^(FULLPERL = .*)/$1 "-Iinc"/m;
	$makefile =~ s/^(PERL = .*)/$1 "-Iinc"/m;
	
	open  MAKEFILE, "> $file" or die "FixMakefile: Couldn't open $file: $!; bailing out";
	print MAKEFILE $makefile or die $!;
	close MAKEFILE or die $!;
}

my $mm = WriteMakefile(%WriteMakefileArgs);
FixMakefile($mm->{FIRST_MAKEFILE} || 'Makefile');
exit(0);

